src/stdlib_stringlist.f90

```diff
@@ -0,0 +1,1063 @@
+!
+!     Note: very preliminary
+!
+!     TODO:
+!     insert( list_end, ... ) in an empty list?
+!     concatenate two string lists
+!
+!     Not implemented yet:
+!     insert a list or an array of character strings
+!     replace a string, list or an array of character strings
+!     concatenate a list with another list or an array
+!
+!     Limited to implemented routines
+!
+module stdlib_stringlist
+    use stdlib_string_type, only: string_type, move, assignment(=) !, char, operator(==)
+    use stdlib_math, only: clip
+    use stdlib_optval, only: optval
+    implicit none
+    private
+
+    public :: stringlist_type, operator(//)
+    public :: list_head, list_tail, fidx, bidx !, operator(+), operator(-)
+
+    type stringlist_index_type
+        private
+        logical :: head
+        integer :: offset
+
+    end type stringlist_index_type
+
+    type(stringlist_index_type), parameter :: list_head      = forward_index(1)
+    ! type(stringlist_index_type), parameter :: list_end       = stringlist_index_type( .false., 0 )
+    ! type(stringlist_index_type), parameter :: list_after_end = stringlist_index_type( .false., 1 )
+    type(stringlist_index_type), parameter :: list_tail      = backward_index(1)
+
+    ! interface operator(+)
+    !     module procedure stringlist_index_add
+    ! end interface
+
+    ! interface operator(-)
+    !     module procedure stringlist_index_subtract
+    ! end interface
+
+    interface fidx
+        module procedure forward_index
+    end interface
+
+    interface bidx
+        module procedure backward_index
+    end interface
+
+    type stringlist_type
+        private
+        integer :: size = 0
+        type(string_type), dimension(:), allocatable :: stringarray
+    
+    contains
+        private
+        procedure, public :: destroy                => destroy_list
+        procedure         :: insert_string_idx      => insert_string_idx_wrap
+        procedure         :: insert_string_int      => insert_string_int_impl
+        procedure         :: insert_stringlist_idx  => insert_stringlist_idx_wrap
+        procedure         :: insert_stringlist_int  => insert_stringlist_int_impl
+        procedure         :: insert_stringarray_idx => insert_stringarray_idx_wrap
+        procedure         :: insert_stringarray_int => insert_stringarray_int_impl
+        generic, public   :: insert                 => insert_string_int,      insert_string_idx,     &
+                                                       insert_stringlist_int,  insert_stringlist_idx, &
+                                                       insert_stringarray_int, insert_stringarray_idx
+        procedure         :: get_string_int         => get_string_int_impl
+        procedure         :: get_string_idx         => get_string_idx_wrap
+        generic, public   :: get                    => get_string_int, get_string_idx
+        procedure, public :: len                    => length_list
```

Comment:
Do you want to bind the `len` function to `stdlib_stringlist(module):stringlist_type(type)`?
Such usage seems to be inconsistent with Fortran's native function `len` and `stdlib_string_type(module):len(interface)`.

Reply:
You mean `size` should have been a better choice to be consistent with Fortran arrays?

Reply:
I am a bit confused, I saw that you bind `len` procedure to `stringlist_type` type (see [stdlib_stringlist.f90#L76](https://github.com/Aman-Godara/stdlib/blob/bc9eda7a9381dbde6c22c8216b37ce447d045e3a/src/stdlib_stringlist.f90#L76)), but then use the `len` function like this (see [stdlib_stringlist.f90#L362](https://github.com/Aman-Godara/stdlib/blob/bc9eda7a9381dbde6c22c8216b37ce447d045e3a/src/stdlib_stringlist.f90#L362)). My point is that `len` should be like this (see [stdlib_string_type.fypp#L60](https://github.com/Aman-Godara/stdlib/blob/bc9eda7a9381dbde6c22c8216b37ce447d045e3a/src/stdlib_string_type.fypp#L60)), it is an independent interface.
The following two syntaxes do not seem to be the same function :
```fortran
length = len(stringlist_type :: list)

---

type(stringlist_type) :: list
length = list%len()
```
So I want to confirm with you : 
> Do you want to bind the `len` function to `stdlib_stringlist(module):stringlist_type(type)`?

Reply:
I should have used `list%len()`. I guess this is object oriented programming stuff where `list` is an object of type `stringlist_type` and `len()` is a method. I apparently have confused this with functional programming throughout the PR. Need to make a lot of changes now.